forem/forem
# Dynamically change number of formatter buttons in md toolbar overflow menu

## What type of PR is this? (check all applicable)


## Description

We had a couple of issues relating to the approach we took to make the toolbar responsive on smaller screens - previously it scrolled horizontally, but there were issues of discoverability and also usability when the scrolling behaviour was triggered on non-touch devices (e.g. in the moderator centre where we render inside a smaller iframe, but also if a user zoomed their viewport for a11y).

Instead of scrolling in small viewports, we now dynamically detect how many items we place in the overflow (3 dots) menu. See the [figma mockups](https://www.figma.com/file/PIPeiiVSADmcpCbiZTdJm9/Editor?node-id=2185%3A8563).

**A small implementation detail:** 

In extremely small screens (super-narrow smartphones like galaxy fold, or mobile devices where a user is choosing to zoom the screen), I've kept the scrolling behaviour as before. I've done this purely as a fall-back for edge cases like screen zooming for a11y, or those rarer width smartphones. 

The solution we have at the moment is fairly lightweight/understandable, and I didn't want to introduce a ton of complexity just for these less-likely scenarios (but also didn't want things to look terrible for those users).

## Related Tickets & Documents



- Closes https://github.com/forem/forem/issues/17714
- Closes https://github.com/forem/forem/issues/18131

## QA Instructions, Screenshots, Recordings

This needs tested in 3 locations:
- The editor
- Comments on a post
- In the mod centre, commenting on a post

In each case, experiment with resizing your screen and check that everything looks sensible 😄  Please also check some things that gave us headaches before:

- In larger screen sizes, tooltips appear on hover
- Regardless of the overflow menu being open, or the bar being horizontally scrollable, there is never a vertical scroll


https://user-images.githubusercontent.com/20773163/181026057-58a36a2c-c72b-401a-a965-7acdcf9ab687.mp4


### UI accessibility concerns?

I think this will make things a bit more usable for all.

## Added/updated tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
- [X] This change does not need to be communicated, and this is why not: small tweak, no new functionality
